build.assets/tooling/cmd/resource-ref-generator/reference/reference.go

```diff
@@ -0,0 +1,332 @@
+	Version string
+	Kind    string
+	resource.ReferenceEntry
+}
+
+// Intended to be executed with a ReferenceContent.
+// Ampersands are replaced with backticks.
+var referenceTemplate string = strings.ReplaceAll(`---
+title: {{ .Resource.SectionName }} Reference
+description: Provides a reference of fields within the {{ .Resource.SectionName }} resource, which you can manage with tctl.
+sidebar_title: {{ .Resource.SectionName }}
+---
+
+{{ if ne .Resource.Kind "" }}
+**Kind**: {{ .Resource.Kind }}
+{{ end }}
+
+{{ if ne .Resource.Version "" }}
+**Version**: {{ .Resource.Version }}
+{{ end }}
+
+{{ .Resource.Description }}
+
+{/*Automatically generated from: {{ .Resource.SourcePath}}*/}
+
+Example:
+
+&&&yaml
+{{ .Resource.YAMLExample -}}
+&&&
+
+{{ if gt (len .Resource.Fields) 0 }}
+|Field Name|Description|Type|
+|---|---|---|
+{{ range .Resource.Fields -}}
+|{{.Name}}|{{.Description}}|{{.Type}}|
+{{ end }} 
+{{ end }}
+
+{{ range .Fields }}
+## {{ .SectionName }}
+
+{{ .Description }}
+
+{/*Automatically generated from: {{ .SourcePath}}*/}
+{{ if ne .YAMLExample "" }}
+Example:
+
+&&&yaml
+{{ .YAMLExample -}}
+&&&
+{{ end }}
+{{ if gt (len .Fields) 0 }}
+|Field Name|Description|Type|
+|---|---|---|
+{{ range .Fields -}}
+|{{.Name}}|{{.Description}}|{{.Type}}|
+{{ end }} 
+{{ end }}
+{{ end }}
+`, "&", "`")
+
+// TypeInfo represents the name and package name of an exported Go type. It
+// makes no guarantees about whether the type was actually declared within the
+// package.
+type TypeInfo struct {
+	// Go package path (not a file path)
+	Package string `yaml:"package"`
+	// Name of the type, e.g., Metadata
+	Name string `yaml:"name"`
+}
+
+// GeneratorConfig is the user-facing configuration for the resource reference
+// generator.
+type GeneratorConfig struct {
+	// Field types that a type must have to be included in the reference. A
+	// type must have one of these field types to be included in the
+	// reference. The fields named here can be embedded fields.
+	RequiredFieldTypes []TypeInfo `yaml:"required_field_types"`
+	// Path to the root of the Go project directory.
+	SourcePath string `yaml:"source"`
+	// Directory where the generator writes reference pages.
+	DestinationDirectory string `yaml:"destination"`
+	// Struct types to exclude from the reference.
+	ExcludedResourceTypes []TypeInfo `yaml:"excluded_resource_types"`
+	// The name of the method that assigns values to the required fields
+	// within a dynamic resource. The generator determines that a type is a
+	// dynamic resource if it has this method.
+	FieldAssignmentMethodName string `yaml:"field_assignment_method"`
+}
+
+// UnmarshalYAML checks that the GeneratorConfig includes all required fields and, if
+// not, returns the first error it encounters.
+func (c GeneratorConfig) UnmarshalYAML(value *yaml.Node) error {
+	if err := value.Decode(&c); err != nil {
+		return fmt.Errorf("could not parse the configuration file as YAML: %v\n", err)
+	}
+
+	switch {
```

Comment:
Nit: It's a little unusual to perform validation like this inside of `UnmarshalYAML` but since this isn't production code, I don't think I'm too worried.